### PR TITLE
Address review comments from #874

### LINKED ISSUE #874: Query on segmentation modeling (deeplabv3)
1. In general, shouldn't the `input size` and `target size` match? Why I need to downscale the target samples for training? For example,  input size 128, target size 32.
2. I got the following warning message. Tested on colab.  

> WARNING:tensorflow:Gradients do not exist for variables ['deep_lab_v3/feature_pyramid/output_P3/kernel:0', 'deep_lab_v3/feature_pyramid/output_P3/bias:0', 'deep_lab_v3/feature_pyramid/output_P4/kernel:0', 'deep_lab_v3/feature_pyramid/output_P4/bias:0', 'deep_lab_v3/feature_pyramid/output_P5/kernel:0', 'deep_lab_v3/feature_pyramid/output_P5/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?

3. `mixed precision` makes loss value to `nan` (probably need to cast the output of segmentation head to `tf.float32`. You might want to introduce `dtype` param in [here](https://github.com/keras-team/keras-cv/blob/673e63cf2fc5b4ff295004a20aac7b5e2544c16d/keras_cv/models/segmentation/__internal__/segmentation_head.py#L51).) 

[Reproducible Code.](https://colab.research.google.com/drive/1x-mB_CPnQkuqSMq7peHWk32h6V9Qi6h3?usp=sharing)

### PR DESCRIPTION
**Pull request by qlzh727**
# What does this PR do?

1. Add a param to scale up the output for segmentation head so that the output mask
can have the same shape as the input image.
2. Enforce the output layer dtype to float32, to avoid the numerical issue when running
with mixed precision.

Fixes #874

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qlzh727** [comment]
@innat

File: keras_cv/models/segmentation/__internal__/segmentation_head.py [hunk 1]
```diff
@@ -32,6 +32,12 @@ class SegmentationHead(layers.Layer):
             to 256.
         activations: str or 'tf.keras.activations', activation functions between the
             conv2D layers and the final classification layer. Default to 'relu'
+        output_scale_factor: int, or a pair of ints, for scale up the output mask.
```
Comment (tanzhenyu) [inline keras_cv/models/segmentation/__internal__/segmentation_head.py:35]: for upsample the output mask.

**Review by tanzhenyu — APPROVED** [review]

File: keras_cv/models/segmentation/__internal__/segmentation_head.py (hunk 1 above)
Comment (qlzh727) [reply keras_cv/models/segmentation/__internal__/segmentation_head.py]: Done.
